### DISCUSSION TITLE
Data Bridge - push a topic+subtopics to broker.emqx.io

### DISCUSSION THREAD

**Discussion by alexdelprete** [emqx/emqx#9078] [Q&A] [answered]
Hi,

first of all thanks for this great project. I just migrated from mosquitto, that has always served me well, very reliable project, but the lack of a GUI has led me to migrate to a more modern broker.

I setup EMQX in an LXC container in proxmox, everything's working as expected, but I had a few issues with the Data Bridge, that I use to replicate a topic + its subtopics to an external public broker (the emqx.io one).

With Mosquitto, I used this configuration for the bridge:

```
# Bridge to EMQX public broker
connection bridge-to-broker
address broker.emqx.io:1883
bridge_protocol_version mqttv50
start_type automatic
cleansession true
try_private false
bridge_attempt_unsubscribe false
notifications false
remote_clientid ha_bridge_weewx
topic wftempest/weewx/# out 0
```

Basically, I replicated the topic **`wftempest/weewx`** with all its subtopics to the public broker.

With EMQX, I created a Data Bridge (MQTT Sink) and a rule to feed it.

Rule:

```
SELECT
  *
FROM
  "wftempest/weewx/#"
```

And the bridge was configured like this:

```
bridges {
  mqtt {
    "weewx-tempest-bridge" {
      connector {
        bridge_mode = false
        clean_start = true
        keepalive = "60s"
        max_inflight = 32
        mode = "cluster_shareload"
        password = ""
        proto_ver = "v5"
        reconnect_interval = "15s"
        replayq {offload = false, seg_bytes = "100MB"}
        retry_interval = "15s"
        server = "broker.emqx.io:1883"
        ssl {
          ciphers = ""
          depth = 10
          enable = false
          reuse_sessions = true
          secure_renegotiate = true
          user_lookup_fun = "emqx_tls_psk:lookup"
          verify = "verify_peer"
          versions = ["tlsv1.3", "tlsv1.2", "tlsv1.1", "tlsv1"]
        }
        username = ""
      }
      direction = "egress"
      enable = true
      payload = ""
      remote_qos = 0
      remote_topic = "wftempest/weewx"
      retain = false
    }
  }
}
```

I used an empty value for the payload because from what I understood, it would have sent everything like I wanted. Instead, on the public broker I was seeing all strange values and messed topics/contents. If I used 
payload = "${payload}" I was only seeing the topic `wftempest/weewx` with the contents of all the subtopics, with no subtopics created with the proper values.

Since I actually needed at least the subtopic `wftempest/weewx/loop` to be working, I found a way to correctly replicate that single one with its content through this configuration (bridge + rule):

```
bridges {
  mqtt {
    "weewx-tempest-bridge" {
      connector {
        bridge_mode = false
        clean_start = true
        keepalive = "60s"
        max_inflight = 32
        mode = "cluster_shareload"
        password = ""
        proto_ver = "v5"
        reconnect_interval = "15s"
        replayq {offload = false, seg_bytes = "100MB"}
        retry_interval = "15s"
        server = "broker.emqx.io:1883"
        ssl {
          ciphers = ""
          depth = 10
          enable = false
          reuse_sessions = true
          secure_renegotiate = true
          user_lookup_fun = "emqx_tls_psk:lookup"
          verify = "verify_peer"
          versions = ["tlsv1.3", "tlsv1.2", "tlsv1.1", "tlsv1"]
        }
        username = ""
      }
      direction = "egress"
      enable = true
      payload = "${payload}"
      remote_qos = 0
      remote_topic = "wftempest/weewx/loop"
      retain = false
    }
  }
}
rule_engine {
  ignore_sys_message = true
  jq_function_default_timeout = "10s"
  rules {
    "rule_wftempest_weewx" {
      actions = ["mqtt:weewx-tempest-bridge"]
      description = ""
      enable = true
      metadata {created_at = 1664370585108}
      sql = "SELECT\n  *\nFROM\n  \"wftempest/weewx/loop\""
    }
  }
}
```

My question is: how can I change the rule (or the bridge) to replicate all the subtopics of `wftempest/weewx`, and not only a specific one?

Thanks for any help on this.

**Comment by lafirest** [comment, marked as answer]
@alexdelprete 
You can use a placeholder in the remote topic, like `remote/${topic}`

**Reply by alexdelprete** [reply]
So I use this rule:

```
SELECT
  *
FROM
  "wftempest/weewx/#"
```

this in the data bridge

```
remote_topic = "${topic}"
```

and `${payload}` in the payload section?

I'll test it immediately...

UPDATE: IT WORKED! :)

Thank you so much @lafirest, I couldn't get it reading the documentation, it was not very clear. Simpler than I thought.

**Reply by lafirest** [reply]
Sorry about this, we will update the document ASAP

**Reply by alexdelprete** [reply]
Absolutely no problem. The support seems great here. Thanks again.

**Comment by alexdelprete** [comment]
Couple of more questions @lafirest:

1. I'm using `broker.emqx.io`, should `bridge_mode = true` in the data bridge config work? It doesn't connect if it's true. What does this setting exactly do?
2. Is `payload = "${payload}"` necessary or can I leave it empty? What's the difference?

Thanks a lot for your help.

**Reply by lafirest** [reply]
@alexdelprete  Hi,
1. As I know the `bridge_mode` only affects the `rap` flag when local subscribe to remote MQTT server
2. `payload = "${payload}"` is not necessary, you can left it empty then you will got whole packet info not only the `payload` data

**Comment by alexdelprete** [comment]
Thanks for the help @lafirest. :)
